### ISSUE TITLE
Refactor walkSqlFiles to use glob

### ISSUE THREAD

**Issue by jonathana** [robconery/massive-js#213]
Looking at the code, there is a comment in walkSqlFiles that says "A recursive directory walker that would love to be refactored".  I'm willing to do that by switching over to using isaacs' glob module.  I've forked into my repo and am working on a branch right now to do that if it makes sense.

One thing I would like to add (which is for my own side purposes) is that in the queryFiles property that is kept of what has been loaded up that holds the name, sql, and path of the file that got loaded up to also add the property path on the object as well (db/foo/bar/somefile.sql becomes foo.bar.somefile).

Is the plan for refactoring walkSqlFiles to use glob solid enough to go forward?  The secondary goal of tracking additional metadata about what endpoints got loaded up is not a hard requirement and I can leave it out.

**Comment by xivSolutions** [comment]
I think we tried using glob for this early on, but moved back to a more traditional function instead. I don't recall what the issue was, but you're welcome to try it on for size and shoot a PR. 

The meta property doesn't sound unreasonable either. However, @robconery generally decides what should go in, and what should be left to the user for this project. My recommendation is, put it together, with appropriate tests, and see what happens.

**Comment by jonathana** [comment]
Regarding the meta property, I think I can finally elucidate a clear reason why I think it should be there.  Right now, a bunch of methods get added to the massive object, especially with a subdirectory tree of queries, on nested properties, and there is no way to determine from the massive object, today, where those properties came from/why.  So if I have massive.subPropA.subPropB.someMethod(), you can't tie it back to the Evaluator that put it on there.  Even worse, you can have massive.subPropA.someMethod() and you can't tie each back to their own Evaluator, and the Evaluators themselves can't differentiate except that the function body is different.

This makes introspection of the resulting massive object difficult, especially if what one wants to do is to either extend or "wrap" certain functionality.  In my case, I'm trying to build a library that uses massive, among other things, to generate JSON data and highlandjs to run a pipeline of transforms on it.  I want to wrap the streaming version of the massive methods that get generated to return a highland stream, but there is no way today to introspect a massive instance with certainty (due to duplicate names) and figure out which methods on it were created by e.g. loading up sql files and turning them into methods on the massive object, because the methods themselves don't indicate why they are on the massive object, and the Evaluator instances don't contain the path information for the methods that got added if you introspect them.

I haven't started the above work of changing around the Evaluator to include this metadata yet, but wanted to explain what I saw in the design.

I think I may know what the issue is with using glob.  The function as constituted walks from the top down, so creating the new methods on the massive object is simplified because when you see a new directory you can create the outer {} to hold the nested functions first.  glob doesn't give you that guarantee, so you have to make sure the hierarchy you need exists when you get a nested function "out of order".

In doing my changes with glob, there are at least 5 ways to work around this:

1. Don't work around it.  Just refactor the existing code using the walk it does today and don't use glob.  I'm pretty sure I can work with that

2. Easy but impactful is to switch from underscore to lodash for its _.get() and _.set() methods that support the nested properties for you.  I'm guessing, however, that that's a library switch massive won't want to make but if it's ok, this is literally a drop-in replacement

3. Easy but requires adding lodash as a dependency anyway: add lodash to the package's dependencies, but only require the 2 functions out of it that are actually needed

4. Pick one of the multiple npm libraries that does the nested object work that would be needed, add that to the dependencies, and require that to do the work.  https://www.npmjs.com/package/nested-objects is an example, suggested only because I know the "author" (he pulled some stuff out of backbone-deep-model to create it) and talked to him about the state of the code, which makes _me_ more comfortable with it anyway

5. Add get() and set() functions or methods somewhere in massive to do the work, either by writing from scratch or copying existing, tested code with a compatible license

I already have a personal git branch with choice 2 done and tests passing.  I'll work on choice 1 now.

**Comment by robconery** [comment]
If I can condense this: 1) you want to refactor the file-walker using glob and 2) you want to restructure the way we track the table info. Let's break this out to separate issues if possible.

If you want to try and refactor the file walker I suppose go for it, however *it works* and I see no reason right now to mess with something that works. I *do* know there's that comment in there but I'm feeling a bit shy about this now.

RE the meta stuff - I'm also pretty shy about restructuring the project to fit your project's needs. I'm all ears on recommended changes based on individual, specific issues you're having, however.

**Comment by jonathana** [comment]
@robconery Thanks for the input, and agreed on the 2 issues.  Before I go off and add more issues, would it be worthwhile to see if we can determine whether either one is worth doing?

I'll cover the metadata thing in this comment, and deal with glob in a second.

I'm not sure about the "restructuring the project to fit your project's needs" bit.  Yes, the change happens to suit something I want to do, but it also cleans up what I think might be a concern about massive as it's structured now.  Today, there are Evaluator objects attached to self.queryFiles, and there are (potentially nested, if you have subdirectories) methods on the massive instance.  However, there is no way, especially if people don't give each SQL file unique names, to identify which method added to the massive instance came from where.  My proposal to extend the Executable with a propertyPath was strictly because while it can be intuited by looking at this.scriptsDir and removing that and everything before it as well as the .sql extension from the Executable.filePath to find what property on the massive instance goes with that Executable.

If being able to introspect the massive instance to determine where the various properties came from isn't a valuable enough use case for massive, then I'm good with that.  I can accomplish the same thing outside of it for my own use cases as described above.

**Comment by jonathana** [comment]
As for the glob thing, I'm willing to give up on that as a proposal.  Using glob would result in less code, but at the cost of adding glob, and at a minimum lodash.get and lodash.set as dependencies (lodash lets you npm install individual methods, so I could get the footprint way down in that regard).  However, in looking at it, it would also allow directory names with dots in them, which will get really screwy because someone could create a directory structure like:
-subA
    -subB
        -runSomeQuery.sql
-subA.subB
    -runSomeQuery.sql

and now there's a name collision, although that could be handled in error handling, but then it puts more code back in.

So I'll drop the offer to refactor to glob, even with the code as it stands now, because I'm not sure glob would be shorter or more reliable.

I can implement the metadata stuff inside the existing code if (and only if) that is something you deem worthwhile.  And if you don't want that, then I'll try and work around it on the outside.

**Comment by robconery** [comment]
Let's close the *glob* discussion as supporting dots in directories is not something I feel like thinking about.

I'm having a hard time dissecting what issue we're focusing on with the file sql. I don't mean to be too direct RE my quip about restructuring to fit your project - but that's exactly the scenario I read after your initial pushback. 

If I'm reading you correctly: you want to know where the methods came from (SQL File, Table, View, etc) - yes? You need this because you're creating a library that does some stuff and you need to know the *provenance* (to use a fancy word) of each method. I think this is an interesting problem to solve - but I don't see how it's a *structural concern* just yet.

My first question would be: can you interrogate `Views`, `Tables` and `Functions` to see if the method exists? Would having a `is_file` function help you out?

**Comment by jonathana** [comment]
Agreed on the glob discussion.

file sql is just a symptom of what I'm trying to get at, so I don't want to dwell on that in particular.  I didn't consider the quip about restructuring as too direct, and I don't want to come off as too direct in anything either.

What I'm trying to know here is: when I look at methods on a massive instance, nested or not, are they something that got added that's doing database access that I might want to wrap?

I'm pretty sure without asking for anything new that I can do that for Table and View stuff without needing anything, although the below might be handy.  The problem really came with the adding of SQL files, especially because they can nest and names will only be unique within a particular nested object.

The simple proposal I'm left with suggesting here regarding metadata is: given an instance of an Executable, let it either specify as a property on itself where on the massive instance the function to execute it got loaded, so that when the code does this to stick a new method on the massive instance from e.g. a SQL file:
```
      rootObject[name] = function () {
        return _exec.invoke.apply(_exec, arguments);
      };
```
I can go back and say "when I find a method on a massive instance, it was put there through the above code?".  I did not see anything obvious when looking at any functions on the massive instance to say "aha, that's doing database access" to work around it.

As I said in a prior comment, I can do it today by interrogating massive.scriptsDir and then massive.queryFiles to get the Executable's and then look at their filePath, however that then creates some implicit dependencies that could break down the road: if massive ever looks for anything other than .sql files, I have to know when that changes.  It would be more reliable if massive gave me a way to get those answers.  Again, whether that's by having a function that looks at Executable.filePath and self.scriptsDir, strips extensions and converts slashes to dots, or whether it just builds the value onto Executable when it's created doesn't matter to me.
